### PR TITLE
Plan 07b: auto-rename rewrite + session-owned frontmatter

### PR DESCRIPTION
**Pull request by maccman**
## What this is

The rename half of [Plan 07](docs/plans/07-backlinks.md): editing a non-daily note's title **automatically rewrites every inbound `[[link]]`** (decided: no confirmation prompt) and preserves the old title as a frontmatter alias. Completes Plan 07 together with #13.

## Session-owned frontmatter (the structural piece)

Probing revealed a pre-existing bug: meowdown reads `---` blocks as thematic-break/setext markdown, so the round-trip guard classified every frontmatter note **lossy → read-only**. The session now owns frontmatter:

- every disk read is split — the editor sees the body only, protection gates on the body, and saves rejoin the **exact original header bytes**;
- `updateFrontmatter(patch)` gives metadata writes (the alias) a channel through the normal save pipeline that never touches the editor view — no reload, no conflict banner, no protected flip;
- `onContent(content, origin)` reports `load` / `external` / `saved` transitions; `saved` is the only user-driven origin, which is exactly what rename detection needs.

## Settled-title detection

[title-rename.ts](apps/desktop/src/editor/title-rename.ts): a rename fires only on a **settled** title — blur, pane teardown, or 5s of quiet after a save — never per keystroke, so intermediate typing states ("My N…") can't spray the graph with rewrites. External content re-baselines without firing (only edits made *here* trigger), pure case changes are not renames (resolution is case-insensitive), and chained renames carry the previous auto-added alias so a re-rename prunes it instead of accreting one alias per edit.

## The rewrite pipeline

[rename.ts](packages/core/src/indexing/rename.ts) (core, DI per conventions §3): **collision guard** — if the old title now resolves to a *different* note, existing links still point somewhere deliberate and are left alone (a stale index resolving to the renamed note itself is not a collision); the renamed note is excluded; writes are serialized and carry the generation token (a rename racing a graph switch dies loudly); a failing source is skipped, not fatal — the alias keeps it resolving. `nextAliases` prunes the prior auto-alias, appends the old title, returns `null` when nothing changes. Rewrites land via `renameWikiLink` (Plan 03 minimal-diff splices, alias-preserving). Progress surfaces in the pane ("Updating links to this note… 3/12").

Dailies are excluded (`trackRenames: !isDaily(path)`) — their date labels are stream chrome, not content. Filenames stay put in the first wave (recorded in the plan doc). The click-side disambiguation picker is explicitly deferred to Plan 08.

## Tests

Tracker semantics (9: quiet timer, re-arm on intermediate titles, settle, revert clears, case-only, external re-baseline, alias chain), pipeline + `nextAliases` (9: rewrite/alias-form preservation, self-exclusion, collision both ways, failure continuation, progress), session frontmatter (5: body-only classification — the read-only bug fix, exact header rejoin, `updateFrontmatter`, clean external header adoption, `onContent` origins), and an end-to-end hook test (heading edit → blur settle → source rewritten + alias recorded). **124 core + 120 desktop + 4 db**; lint/typecheck clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **High Risk**
> Touches save/conflict/frontmatter semantics and performs graph-wide link rewrites with generation-pinned writes; mistakes could corrupt notes or strand links, though collision guards and broad tests mitigate this.
> 
> **Overview**
> Delivers **Plan 07b**: when a non-daily note’s title **settles** (blur, pane close, or ~5s after save), the app rewrites inbound `[[links]]` and stores the old title as a frontmatter alias—without prompting and without firing on every keystroke.
> 
> **Session-owned frontmatter** fixes notes with `---` blocks wrongly opening read-only: the session splits header/body on load, the editor edits body only, saves rejoin exact header bytes, and `updateFrontmatter` / `onContent` support alias writes without reloading the editor.
> 
> **Rename stack**: `title-rename` tracker + `rename-coordinator` (serialized rewrites, conflict gating, alias via live session or disk) + core `rewriteLinksForTitleChange` / `getLinkSources` / `nextAliases` (collision guard, code-block-safe splices). Dailies opt out via `trackRenames: !isDaily(path)`; the pane shows rewrite progress.
> 
> Extensive unit and hook tests cover frontmatter, collisions, failed rewrites, note switches, quit flush, and edge cases; the backlinks plan doc is updated for 07b contracts.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9b8ed8089d2b5c57ba2ac32e715af2cfe5f978b8. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Automatic title-rename coordination with optional in-pane tracking and a visible rename progress indicator
  * In-editor frontmatter patching that updates headers without reloading the editor
  * Exposed full-note content access for session workflows (used for load/save/external events)

* **Improvements**
  * Safe, ordered rename rewrites across linked notes with collision handling and alias preservation
  * Preserves exact frontmatter bytes and treats header/body separately during saves

* **Tests**
  * Expanded coverage for rename flows, frontmatter ownership, collisions, and partial-failure cases

* **Documentation**
  * Updated backlinks plan with delivery split and rename/link-rewrite contracts

### CODE REVIEW

File: apps/desktop/src/editor/use-note-document.ts [hunk 1]
```diff
@@ -69,6 +95,46 @@ export function useNoteDocument(
...
+            },
+            onProgress: (done, total) => setRenameProgress({ done, total }),
+          })
+          const aliases = nextAliases(
+            parseNote({ path, source: rename.content }).frontmatter.aliases,
+            rename,
+          )
+          if (aliases !== null) {
+            sessionRef.current?.updateFrontmatter({ aliases })
+          }
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/use-note-document.ts]: Fixed in debdfe3 — the rename closes over its own effect's session instead of sessionRef: while the pane is open the alias goes through that session's frontmatter channel (then flushes), and after teardown it's written directly to disk with the generation token (no editor to disturb; a reopened pane reconciles it as an external change). Regression test covers the teardown path.

File: apps/desktop/src/editor/use-note-document.ts [hunk 2]
```diff
@@ -133,7 +220,12 @@ export function useNoteDocument(
     if (!path) {
       return
     }
-    const flush = (): void => void sessionRef.current?.flush()
+    const flush = (): void => {
+      const settled = sessionRef.current?.flush()
+      // Blur is a settle point for title renames — but only after the flushed
+      // save lands, so the tracker has seen the final title.
+      void settled?.then(() => trackerRef.current?.settle())
+    }
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/use-note-document.ts:208]: Fixed in debdfe3 — the quit-flush registry entry settles the tracker after the buffer flush and awaits the rename chain, so a pending title change rewrites links and records its alias before exit; covered by a flushAllNotes regression test.

File: apps/desktop/src/editor/title-rename.ts [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    if (foldKey(title) === foldKey(baselineTitle)) {
+      // Same key: resolution is case-insensitive, so a pure case tweak is not
+      // a rename — and a reverted edit clears whatever was pending.
+      cancelTimer()
+      pending = null
+      return
+    }
+    pending = { title, content }
+    cancelTimer()
+    timer = setTimeout(fire, quietMs)
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/title-rename.ts:144]: Behavior is correct and now pinned in 913a727 — with an explicit `title:`, the H1 isn't the note's title (deriveTitle precedence, identical in tracker and indexer), links resolve against `title:` regardless, so there's nothing to rewrite. The real limitation — frontmatter-titled notes can't be renamed from the editor yet — is recorded in the plan doc for a later title-field surface.

File: apps/desktop/src/editor/use-note-document.ts [hunk 4]
```diff
@@ -69,6 +96,63 @@ export function useNoteDocument(
...
+              if (patched !== content) {
+                await writeNote(path, patched, generation)
+              }
+            }
+          }
+        } catch (cause) {
+          console.error('rename link rewrite failed:', cause)
+        } finally {
+          setRenameProgress(null)
+        }
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/use-note-document.ts]: Fixed in 913a727 — the tracker gains a canFire gate ('no conflict parked') checked before any state mutation, so a settled rename during a conflict stays pending instead of rewriting the graph for a contested title: 'keep mine' re-saves and re-arms it, 'load theirs' re-baselines and cancels it. Covered by a tracker test.

File: apps/desktop/src/editor/use-note-document.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/use-note-document.ts]: Fixed in debdfe3 (binding the alias to the owning session) and now structurally eliminated in 0dea163 — the rename lifecycle is extracted to rename-coordinator.ts and never touches a React ref, so it can't observe another note's session regardless of timing. Added the requested regression test: a rename settling after a switch to note B rewrites sources and lands A's alias on disk while B stays untouched.

File: apps/desktop/src/editor/use-note-document.ts [hunk 5]
```diff
@@ -69,6 +98,73 @@ export function useNoteDocument(
...
+              // moment to persist, and quit-time teardown awaits this chain.
+              await session.flush()
+            } else {
+              const content = await readNote(path)
+              const patched = upsertFrontmatter(content, { aliases })
+              if (patched !== content) {
+                await writeNote(path, patched, generation)
+              }
+            }
+          }
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/use-note-document.ts]: Fixed in 81a49a7 — on collision the coordinator now skips alias placement too: the old title belongs to another note, so recording it as our alias would plant a competing key that never wins while that note exists and silently re-points links if it's deleted. End-to-end test pins both halves (no rewrite, no alias).

File: apps/desktop/src/editor/use-note-document.ts [hunk 6]
```diff
@@ -133,9 +183,19 @@ export function useNoteDocument(
     if (!path) {
       return
     }
-    const flush = (): void => void sessionRef.current?.flush()
+    const flush = (): void => {
+      const settled = sessionRef.current?.flush()
+      // Blur is a settle point for title renames — but only after the flushed
+      // save lands, so the tracker has seen the final title.
+      void settled?.then(() => coordinatorRef.current?.settle())
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/use-note-document.ts]: Fixed in 21a40c0 — the blur handler (and the quit registry entry, which had the same race) now captures the session/coordinator pair at event time instead of re-reading the refs inside the flush continuation, so a mid-flush note switch can't settle the new note's coordinator.

File: apps/desktop/src/editor/title-rename.ts [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+      to: pending.title,
+      previousAutoAlias,
+      content: pending.content,
+    }
+    // The title we just renamed away from becomes this session's auto-alias —
+    // a follow-up rename prunes it instead of accreting one alias per edit.
+    previousAutoAlias = baselineTitle
+    baselineTitle = pending.title
+    pending = null
+    onRename(rename)
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/title-rename.ts:102]: Fixed in 21a40c0 — baseline-at-fire is deliberate (re-arming on failure would re-fire with a stale from after further edits), but you're right the thrown-rewrite path also skipped the alias, which is the designed safety net. Alias placement now runs even when the rewrite errors (only a collision skips it), and the session-path alias retries through the normal save pipeline; test pins failed-rewrite-still-aliases.

File: apps/desktop/src/editor/title-rename.ts [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+      // Same key: resolution is case-insensitive, so a pure case tweak is not
+      // a rename — and a reverted edit clears whatever was pending.
+      cancelTimer()
+      pending = null
+      return
+    }
+    pending = { title, content }
+    cancelTimer()
+    timer = setTimeout(fire, quietMs)
+  }
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/title-rename.ts:145]: Fixed in ae79511 — titles now count only when the content declares one (frontmatter title: or an H1); parseNote's filename-stem fallback no longer baselines, so a lazy note's first heading is a birth, not a rename (no rewrite, no filename alias). Pinned by tracker tests and a lazy-note end-to-end.

File: apps/desktop/src/editor/use-note-document.ts [hunk 9]
```diff
@@ -69,6 +87,19 @@ export function useNoteDocument(
...
+    // *this* effect's session, never to a ref another note could repoint.
+    const coordinator = trackRenames
+      ? createRenameCoordinator({
+          path,
+          generation: () => generationRef.current,
+          canFire: () => conflictRef.current === null,
+          onProgress: setRenameProgress,
+        })
+      : null
+    coordinatorRef.current = coordinator
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/use-note-document.ts:109]: Fixed in ae79511 — onProgress is gated on the owning effect being live and renameProgress resets on cleanup, so a rewrite finishing after a note switch can't paint its status over the newly opened note.

File: apps/desktop/src/editor/rename-coordinator.ts [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+          rename,
+        )
+        if (aliases !== null) {
+          if (!paneClosed && session !== null) {
+            // Through the owning session's frontmatter channel — the editor
+            // view never churns — and flushed rather than riding the debounce:
+            // a settle is exactly the moment to persist, and quit-time
+            // teardown awaits this chain.
+            session.updateFrontmatter({ aliases })
+            await session.flush()
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/rename-coordinator.ts]: Fixed in b7deea6 — alias placement now computes against the note's current frontmatter at placement time (the session's live content on the open path, read+patched in the same tick; the freshly-read disk content on the fallback path) instead of the settle snapshot, so concurrently-gained aliases survive the whole-key replace. Test pins it.

File: docs/plans/07-backlinks.md [hunk 11]
```diff
@@ -115,7 +129,11 @@ link UX (create-from-unresolved largely subsumes the failure case).
 
 - **Rename rewrite correctness** (links in code blocks, partial matches, ambiguous
   titles). Mitigate: AST-based edits only (Plan 03), skip code contexts, require
-  disambiguation for collisions, batch + checkpoint.
+  disambiguation for collisions, batch + checkpoint. *(Accepted edge, 07b: the
+  collision guard reads the index, which lags the watcher debounce — a note
+  created with the old title inside that sub-second window can be missed.
+  Resolution stays deterministic and the old-title alias still lands, so links
+  keep resolving; the late-created note wins future resolutions.)*
```
[the remark this answers is not in the document]
Comment (maccman) [reply docs/plans/07-backlinks.md]: Fixed in 7a1f6b6 — the risks line now describes the delivered behavior (collision guard: skip rewrite + alias, deterministic first-wave resolution, picker deferred to Plan 08) instead of the pre-07b 'require disambiguation' wording; also added the renameProgress-isolation assertion from the nitpick.

File: apps/desktop/src/editor/rename-coordinator.ts [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+
+  // Rewrite inbound links across the graph, then record the old title as an
+  // alias on this note. Every write carries the generation read at run time
+  // (stale → loud rejection in Rust — a rename pending across a graph switch
+  // is dropped, never cross-written).
+  const runRename = (rename: TitleRename): void => {
+    chain = chain.then(async () => {
+      const gen = generation()
+      if (gen === null) {
+        return
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/rename-coordinator.ts:80]: Addressed in 720995c — the path is unreachable in the current wiring (a rename only arms after a landed save, which requires a generation; NotePane mounts only under an open graph; an unmounted pane's ref keeps its frozen non-null generation, and Rust rejects staleness loudly), but since the baseline has already advanced by fire time, the drop now logs a loud diagnosable error instead of silently no-opping for any future caller.

File: apps/desktop/src/editor/rename-coordinator.ts [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+        } else {
+          // Pane gone: write directly to disk. No editor left to disturb;
+          // a reopened pane reconciles it like any external change.
+          const content = await readNote(path)
+          const aliases = nextAliases(aliasesOf(content), rename)
+          if (aliases !== null) {
+            const patched = upsertFrontmatter(content, { aliases })
+            if (patched !== content) {
+              await writeNote(path, patched, gen)
+            }
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/rename-coordinator.ts:148]: Fixed in 994d9fb — a session registry (path → live session) lets the closed coordinator discover that the note was reopened and place the alias through the live session's frontmatter channel instead of the disk (no watcher race, no self-inflicted conflict, 'keep mine' keeps the alias); the disk write remains only for paths with no session that can take the patch. End-to-end test covers rename → reopen + dirty edit → no conflict, edit and alias both persist.

File: apps/desktop/src/editor/rename-coordinator.ts [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+          // the session. Through its frontmatter channel — the editor view
+          // never churns — and flushed rather than riding the debounce: a
+          // settle is exactly the moment to persist, and quit-time teardown
+          // awaits this chain.
+          const aliases = nextAliases(aliasesOf(owner.content()), rename)
+          placed = aliases === null || owner.updateFrontmatter({ aliases })
+          if (placed && aliases !== null) {
+            await owner.flush()
+          }
+        }
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/rename-coordinator.ts:136]: Behavior is correct as designed, pinned in 9b8ed80 — under a parked conflict the alias rides the session's in-memory header (saves pause; the patch isn't dropped) and lands with 'keep mine'; 'load theirs' discarding it is the user explicitly choosing external content over the rename's consequences. A disk fallback here would write under a parked conflict — clobbering the very 'theirs' the conflict model exists to protect.
